File: examples/go/clients/advanced/hooks.go
package main

import (
	"context"
	"fmt"
	"net/http"

	x402 "github.com/coinbase/x402/go"
	x402http "github.com/coinbase/x402/go/http"
	evm "github.com/coinbase/x402/go/mechanisms/evm/exact/client"
	evmsigners "github.com/coinbase/x402/go/signers/evm"
)

/**
 * Hooks Example
 *
 * This demonstrates how to register hooks for payment creation lifecycle events.
 * Hooks allow you to add custom logic at different stages:
 * - OnBeforePaymentCreation: Called before payment creation starts, can abort
 * - OnAfterPaymentCreation: Called after successful payment creation
 * - OnPaymentCreationFailure: Called when payment creation fails, can recover
 *
 * This is an advanced feature useful for:
 * - Logging payment events for debugging and monitoring
 * - Custom validation before allowing payments
 * - Error recovery strategies
 * - Metrics and analytics collection
 */

func runHooksExample(ctx context.Context, evmPrivateKey, url string) error {
	fmt.Println("🔧 Creating client with payment lifecycle hooks...\n")

	// Create signer from private key
	evmSigner, err := evmsigners.NewClientSignerFromPrivateKey(evmPrivateKey)
	if err != nil {
		return err
	}

	// Create client with scheme registration
	client := x402.Newx402Client().
		Register("eip155:*", evm.NewExactEvmScheme(evmSigner, nil))

	// Register lifecycle hooks

	// OnBeforePaymentCreation: Called before payment is created
	// Use this for logging, validation, or aborting payment creation
	client.OnBeforePaymentCreation(func(ctx x402.PaymentCreationContext) (*x402.BeforePaymentCreationHookResult, error) {
		fmt.Printf("🔍 [BeforePaymentCreation] Creating payment for:\n")
		fmt.Printf("   Network: %s\n", ctx.SelectedRequirements.GetNetwork())
		fmt.Printf("   Scheme: %s\n", ctx.SelectedRequirements.GetScheme())
		fmt.Println()

		// You can abort payment creation by returning:
		// return &x402.BeforePaymentCreationHookResult{
		//     Abort: true,
		//     Reason: "Payment not allowed for this resource",
		// }, nil

		return nil, nil // Continue with payment creation
	})

	// OnAfterPaymentCreation: Called after payment is successfully created
	// Use this for logging, metrics, or other side effects
	client.OnAfterPaymentCreation(func(ctx x402.PaymentCreatedContext) error {
		fmt.Printf("✅ [AfterPaymentCreation] Payment created successfully\n")
		fmt.Printf("   Version: %d\n", ctx.Version)
		fmt.Println()

		// Perform side effects like logging to database, sending metrics, etc.
		// Errors here are logged but don't fail the payment

		return nil
	})

	// OnPaymentCreationFailure: Called when payment creation fails
	// Use this for error recovery or alternative payment methods
	client.OnPaymentCreationFailure(func(ctx x402.PaymentCreationFailureContext) (*x402.PaymentCreationFailureHookResult, error) {
		fmt.Printf("❌ [OnPaymentCreationFailure] Payment creation failed: %v\n", ctx.Error)
		fmt.Println()

		// You could attempt to recover by providing an alternative payload:
		// return &x402.PaymentCreationFailureHookResult{
		//     Recovered: true,
		//     Payload: alternativePayload,
		// }, nil

		return nil, nil // Don't recover, let it fail
	})

	// Create HTTP client wrapper
	httpClient := x402http.Newx402HTTPClient(client)
	wrappedClient := x402http.WrapHTTPClientWithPayment(http.DefaultClient, httpClient)

	// Make request to trigger hooks
	fmt.Printf("🌐 Making request to: %s\n\n", url)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return err
	}

	resp, err := wrappedClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	fmt.Println("✅ Request completed successfully with hooks\n")

	if err := printResponse(resp, "Response with hooks"); err != nil {
		return err
	}
	printPaymentDetails(resp.Header)
	return nil
}

File: examples/typescript/clients/advanced/hooks.ts
import { privateKeyToAccount } from "viem/accounts";
import { x402Client } from "@x402/fetch";
import { ExactEvmScheme } from "@x402/evm/exact/client";
import { UptoEvmScheme } from "@x402/evm/upto/client";

/**
 * Hooks Example
 *
 * This demonstrates how to register hooks for payment creation lifecycle events.
 * Hooks allow you to add custom logic at different stages:
 * - onBeforePaymentCreation: Called before payment creation starts, can abort
 * - onAfterPaymentCreation: Called after successful payment creation
 * - onPaymentCreationFailure: Called when payment creation fails, can recover
 *
 * This is an advanced feature useful for:
 * - Logging payment events for debugging and monitoring
 * - Custom validation before allowing payments
 * - Error recovery strategies
 * - Metrics and analytics collection
 *
 * @param evmPrivateKey - The EVM private key for signing
 * @param url - The URL to make the request to
 */
export async function runHooksExample(evmPrivateKey: `0x${string}`, url: string): Promise<void> {
  console.log("🔧 Creating client with payment lifecycle hooks...\n");

  const evmSigner = privateKeyToAccount(evmPrivateKey);

  const client = new x402Client()
    .register("eip155:*", new ExactEvmScheme(evmSigner))
    .register("eip155:*", new UptoEvmScheme(evmSigner))
    .onBeforePaymentCreation(async context => {
      console.log("🔍 [BeforePaymentCreation] Creating payment for:");
      console.log(`   Network: ${context.selectedRequirements.network}`);
      console.log(`   Scheme: ${context.selectedRequirements.scheme}`);
      console.log();

      // You can abort payment creation by returning:
      // return { abort: true, reason: "Payment not allowed for this resource" };
    })
    .onAfterPaymentCreation(async context => {
      console.log("✅ [AfterPaymentCreation] Payment created successfully");
      console.log(`   Version: ${context.paymentPayload.x402Version}`);
      console.log();

      // Perform side effects like logging to database, sending metrics, etc.
      // Errors here are logged but don't fail the payment
    })
    .onPaymentCreationFailure(async context => {
      console.log(`❌ [OnPaymentCreationFailure] Payment creation failed: ${context.error}`);
      console.log();

      // You could attempt to recover by providing an alternative payload:
      // return { recovered: true, payload: alternativePayload };
    });

  const { wrapFetchWithPayment } = await import("@x402/fetch");
  const fetchWithPayment = wrapFetchWithPayment(fetch, client);

  console.log(`🌐 Making request to: ${url}\n`);
  const response = await fetchWithPayment(url, { method: "GET" });
  const body = await response.json();

  console.log("✅ Request completed successfully with hooks\n");
  console.log("Response body:", body);

  // Extract payment response from headers
  const { x402HTTPClient } = await import("@x402/fetch");
  const paymentResponse = new x402HTTPClient(client).getPaymentSettleResponse(name =>
    response.headers.get(name),
  );
  if (paymentResponse) {
    console.log("\n💰 Payment Details:", paymentResponse);
  }
}
